### PR TITLE
Fix: Migrate TextField to TS

### LINKED ISSUE #19127: Migrate components to TS: TextField
### Description
Migrate `TextField` component  in `ui/components/component-library` to TypeScript

Dependencies
- https://github.com/MetaMask/metamask-extension/issues/19239
- https://github.com/metamask/metamask-extension/issues/19122

### Technical Details

- Convert Objects in `SNAKE_CASE` to enums in `PascalCase` (name and keys)
- Ensure enums use singular not plural e.g. `Size` not `Sizes`
- Update file name from `component-name.contants.js` to `component-name.types.ts` and add types
- Update file name from `component-name.js` to `component-name.tsx` and add types
- Update file name from `component-name.test.js` to `component-name.test.tsx`
- Update file name from `component-name.stories.js` to `component-name.stories.tsx`
- Update all instances of component objects to enums
- Update all documentation

### Acceptance Criteria

- Component has been converted to TypeScript
- Stories, documentation and tests have been updated inside of the `component-name` folder
- Before/After screen cast of component stories in storybook to show no visual regression after migration

### Difficulty: Intermediate

Good first issue for: External contributors who are familiar with TypeScript and want to contribute to improving the stability of the extension

### PR DESCRIPTION
**Pull request by dhruvv173**
## Explanation
This PR focuses on the migration of the `TextField` component, currently in JavaScript, from the component-library folder to TypeScript. As part of this process, the JavaScript version is now deprecated. 

To minimize disruption to the application, we have updated all import paths from the JavaScript version to point to the deprecated version. 

Following the approval and merge of this PR, we will proceed to replace every instance of the deprecated JavaScript component with the new TypeScript version. We will ensure that all functionalities are working as expected after this transition.

* Fixes #19127 

### Before

https://github.com/MetaMask/metamask-extension/assets/8112138/3e61fe03-bf41-48e5-8e9f-4618df7a5b3a

### After

https://github.com/MetaMask/metamask-extension/assets/8112138/4523091c-7826-471a-8b3f-951b991908c1

## Manual Testing Steps

- Go to the storybook build of this PR
- Search `TextField` in the search bar and go to the component-library version
- Check documentation, stories and controls 

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes

## Pre-merge reviewer checklist

- [x] Sufficient automated test coverage has been added
- [x] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [x] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: ui/components/component-library/text-field/text-field.types.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+   */
+  onChange?: () => void;
+  /**
+   * Callback fired when the TextField is clicked on
+   */
+  onClick?: () => void;
+  /**
+   * Callback fired on focus
+   */
+  onFocus?: () => void;
```
Comment (dhruvv173) [inline ui/components/component-library/text-field/text-field.types.ts:95]: it would be better to add typesafety here?

File: ui/components/component-library/text-field/text-field.test.tsx [hunk 2]
```diff
@@ -101,11 +99,11 @@ describe('TextField', () => {
       <>
         <TextField inputProps={{ 'data-testid': 'text-field-text' }} />
         <TextField
-          type="number"
+          type={TextFieldType.Number}
           inputProps={{ 'data-testid': 'text-field-number' }}
```
Comment (dhruvv173) [inline ui/components/component-library/text-field/text-field.test.tsx:103]: looks like we'll have to add `data-testid` to `InputProps` interface(?)

File: ui/components/component-library/text-field/text-field.types.ts (hunk 1 above)
Comment (georgewrmarshall) [reply ui/components/component-library/text-field/text-field.types.ts:95]: Agreed! Is there specific types for these event handlers that we can add?

File: ui/components/component-library/text-field/text-field.types.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  id?: string;
+  /**
+   * The the component that is rendered as the input
+   * Defaults to the Text component
+   */
+  InputComponent?: InputComponent;
+  /**
+   * Attributes applied to the `input` element.
+   */
+  inputProps?: InputProps<'span'>;
```
[the remark this answers is not in the document]
Comment (dhruvv173) [reply ui/components/component-library/text-field/text-field.types.ts:57]: oh yes, you're right. My bad.

File: ui/components/component-library/index.js [hunk 4]
```diff
@@ -32,7 +32,7 @@ export { Tag } from './tag';
 export { TagUrl } from './tag-url';
 export { Text, ValidTag, TextDirection, InvisibleCharacter } from './text';
 export { Input, InputType } from './input';
-export { TextField, TEXT_FIELD_TYPES, TEXT_FIELD_SIZES } from './text-field';
```
Comment (georgewrmarshall) [inline ui/components/component-library/index.js:35]: Because `TEXT_FIELD_TYPES` and `TEXT_FIELD_SIZES ` constants are used throughout the codebase I would suggest deprecating them instead of removing them. Below is doing a search for the constants you can see that there are quite a few results

Screenshot 2023-08-16 at 10 46 36 AM

That way we don't have to replace every instance in the code base but we are still notifying engineers to cease using them. Here is an example of how we deprecated the `DISPLAY` constant in favor of the `Display` enum. 

https://github.com/MetaMask/metamask-extension/blob/6512cacec837b79703e7a76c949d05f1b0a4f722/ui/helpers/constants/design-system.ts#L324-L340

Instead of link an issue here we can just keep it to:  Help to replace `TEXT_FIELD_TYPES ` with `TextFieldType` by submitting a PR

**Review by georgewrmarshall — CHANGES_REQUESTED** [review]
Hey @dhruvv173, this is looking great! I need to do a full review and checkout your branch but have left some initial suggestions

File: ui/components/component-library/index.js (hunk 4 above)
Comment (dhruvv173) [reply ui/components/component-library/index.js:35]: yes, this sounds much better. Will do it.

File: ui/components/component-library/text-field/text-field.types.ts (hunk 1 above)
Comment (dhruvv173) [reply ui/components/component-library/text-field/text-field.types.ts:95]: we can do this?
```
onBlur?: (event: React.FocusEvent<HTMLInputElement>) => void;
onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
onClick?: (event: React.MouseEvent<HTMLInputElement>) => void;
onFocus?: (event: React.FocusEvent<HTMLInputElement>) => void;
```

File: ui/components/component-library/index.js (hunk 4 above)
Comment (dhruvv173) [reply ui/components/component-library/index.js:35]: done [`here`](https://github.com/MetaMask/metamask-extension/pull/20477/commits/4318f99409d540d47fb69ebc5cb0f38524469733)

after screenshots:
image
image

Comment (georgewrmarshall) [reply ui/components/component-library/index.js:35]: Excellent!

File: ui/components/component-library/text-field/README.mdx [hunk 5]
```diff
@@ -1,4 +1,4 @@
-import { Story, Canvas, ArgsTable } from '@storybook/addon-docs';
+import { Story, Canvas, ArgTypes } from '@storybook/addon-docs';
```
Comment (georgewrmarshall) [inline ui/components/component-library/text-field/README.mdx:1]: `ArgsTable` has been deprecated in favor of `ArgTypes` there are a few more but we can address them in a storybook clean up

**Comment by georgewrmarshall** [comment]
Hey @dhruvv173, I think this is pretty close. We just need to fix the linting errors. One for the ref and one to add all of the previous TextField PropTypes to FormTextField. Do you have time to give those a go?

File: ui/components/component-library/form-text-field/form-text-field.js
Comment (dhruvv173) [inline ui/components/component-library/form-text-field/form-text-field.js:1]: https://github.com/MetaMask/metamask-extension/blob/8737068f5e400fac4d89512643d68a18de6ccf8f/ui/components/component-library/form-text-field/form-text-field.js#L44

we can change this to `type = TextFieldType.Text` ?

**Review by georgewrmarshall — COMMENTED** [review]
Hey @dhruvv173, this is starting to touch a lot of files because of the `TextField` dependency in other files. I think we may need to approach this the same way we have done with other components during TypeScript migrations where we move the JS component into a deprecated folder within the component folder and update the import paths. Here is an example of the `Icon` component migration PR https://github.com/MetaMask/metamask-extension/pull/18431/files#diff-61407b67e861a05c5665c4230fb2d705dede5b6b97c50a2b9b7cbf4360cb5084. Would you be able to implement this here?

**Comment by dhruvv173** [comment]
> Hey @dhruvv173, this is starting to touch a lot of files because of the `TextField` dependency in other files. I think we may need to approach this the same way we have done with other components during TypeScript migrations where we move the JS component into a deprecated folder within the component folder and update the import paths. Here is an example of the `Icon` component migration PR https://github.com/MetaMask/metamask-extension/pull/18431/files#diff-61407b67e861a05c5665c4230fb2d705dede5b6b97c50a2b9b7cbf4360cb5084. Would you be able to implement this here?

yes I'll check the `Icon` component PR and try to implement the same

File: ui/components/component-library/text-field/deprecated/text-field.js [hunk 6]
```diff
@@ -8,11 +8,11 @@ import {
...
   BorderRadius,
   BackgroundColor,
-} from '../../../helpers/constants/design-system';
+} from '../../../../helpers/constants/design-system';
 
-import Box from '../../ui/box';
+import Box from '../../../ui/box';
 
-import { Input } from '../input';
+import { Input } from '../../input';
```
Comment (dhruvv173) [inline ui/components/component-library/text-field/deprecated/text-field.js:15]: updating imports

**Comment by dhruvv173** [comment]
hey @georgewrmarshall for the components using deprecated `TextField` component, how do we want to replace them? directly with the updated TS component or the JS one?
https://github.com/MetaMask/metamask-extension/blob/973e9cebc2ed4dfd964b217b75c34f30e2e7b09d/ui/pages/settings/contact-list-tab/add-contact/add-contact.component.js#L5

File: ui/components/component-library/form-text-field/form-text-field.js [hunk 7]
```diff
@@ -10,7 +10,7 @@ import {
 
 import Box from '../../ui/box/box';
 
-import { TextField } from '../text-field';
+import { TextField, TextFieldSize, TextFieldType } from '../text-field';
```
Comment (georgewrmarshall) [inline ui/components/component-library/form-text-field/form-text-field.js:13]: Can we use the deprecated version here as well and revert all the propType changes? So the only changes to this file will be the import path of the TextField e.g.
```
import { TextField } from '../text-field/deprecated';
```

File: ui/components/component-library/text-field-search/text-field-search.js [hunk 8]
```diff
@@ -100,10 +100,6 @@ TextFieldSearch.propTypes = {
    * Attributes applied to the `input` element.
    */
   inputProps: PropTypes.object,
-  /**
-   * FormTextField accepts all the props from TextField and Box
-   */
-  ...TextField.propTypes,
```
Comment (georgewrmarshall) [inline ui/components/component-library/text-field-search/text-field-search.js:106]: Now that we are using the deprecated JS version here I think we can revert these changes to the propTypes here. So the only changes to this file will be the import path of the TextField e.g.
```
import { TextField, TEXT_FIELD_TYPES } from '../text-field/deprecated';
```

**Review by georgewrmarshall — COMMENTED** [review]
Hey @dhruvv173, this is looking really close I've left some small suggestions. I'm working on solving the ref function it's tricky! But once these changes are in and we can solve the ref function I think this will be in a really good place

File: ui/components/component-library/form-text-field/form-text-field.js
Comment (georgewrmarshall) [reply ui/components/component-library/form-text-field/form-text-field.js:1]: Eventually, yes! Let's do this in the migration of the FormTextField component so we can keep the changes in this PR to a minimum

File: ui/components/component-library/text-field/text-field.types.ts [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+/**
+ * @deprecated `TEXT_FIELD_SIZES` const has been deprecated in favor of the `TextFieldSize` enum which can still be imported from `ui/components/component-library/index.js`
+ *
+ * Help to replace `TEXT_FIELD_SIZES` with `TextFieldSize`
+ */
+export const TEXT_FIELD_SIZES = {
+  SM: Size.SM,
+  MD: Size.MD,
+  LG: Size.LG,
+};
```
Comment (georgewrmarshall) [inline ui/components/component-library/text-field/text-field.types.ts:25]: I think we can remove these now that we have deprecated the entire JS component and updated the import statements for all the instances in the codebase

```suggestion
```

File: ui/components/component-library/text-field/text-field.types.ts [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+ * @deprecated `TEXT_FIELD_TYPES` const has been deprecated in favor of the `TextFieldType` enum which can still be imported from `ui/components/component-library/index.js`
+ *
+ * Help to replace `TEXT_FIELD_TYPES` with `TextFieldType`
+ */
+export const TEXT_FIELD_TYPES = {
+  TEXT: 'text',
+  NUMBER: 'number',
+  PASSWORD: 'password',
+  SEARCH: 'search',
+};
```
Comment (georgewrmarshall) [inline ui/components/component-library/text-field/text-field.types.ts:44]: Same here
```suggestion
```

**Review by georgewrmarshall — COMMENTED** [review]
Still need to do a full review but noticed that we have some redundancies in the `.types.ts` file

File: ui/components/app/snaps/keyring-snap-removal-warning/keyring-snap-removal-warning.tsx [hunk 11]
```diff
@@ -156,7 +157,7 @@ export default function KeyringRemovalSnapWarning({
                   inputProps={{
                     'data-testid': 'remove-snap-confirmation-input',
                   }}
-                  type="text"
+                  type={TextFieldType.Text}
```
Comment (georgewrmarshall) [inline ui/components/app/snaps/keyring-snap-removal-warning/keyring-snap-removal-warning.tsx:160]: Because we are changing the architecture of this component quite substantially let's use the deprecated JS version for all instances. Then replace them one by one so we can check that all of the inputs still work.

File: ui/components/component-library/text-field/text-field.types.ts [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+import { Size } from '../../../helpers/constants/design-system';
+import { InputProps } from '../input';
+import { InputComponent } from '../input/input.types';
+
+export enum TextFieldSize {
+  Sm = 'sm',
+  Md = 'md',
+  Lg = 'lg',
+}
+
```
Comment (georgewrmarshall) [inline ui/components/component-library/text-field/text-field.types.ts:15]: nit: remove extra line break
```suggestion
```

File: ui/components/ui/update-nickname-popover/update-nickname-popover.js [hunk 13]
```diff
@@ -103,9 +103,9 @@ export default function UpdateNicknamePopover({
...
           onChange={handleMemoChange}
           placeholder={t('addMemo')}
-          fullWidth
+          width={BlockSize.Full}
           margin="dense"
-          multiline
+          // multiline (this is from MUI component) To Do: update TextField component to support multiline/textarea input
           rows={3}
           classes={{
             inputMultiline: 'update-nickname__content__text-area',
```
Comment (georgewrmarshall) [inline ui/components/ui/update-nickname-popover/update-nickname-popover.js:111]: Let's use the deprecated version here and update the import paths

**Review by georgewrmarshall — CHANGES_REQUESTED** [review]
Let's use the deprecated JS version for all instances and update the import paths. This will prevent any bugs getting introduced from the rearchitecting and save us having to check those instances. Then we can replace them one by one

File: ui/components/app/snaps/keyring-snap-removal-warning/keyring-snap-removal-warning.tsx (hunk 11 above)
Comment (dhruvv173) [reply ui/components/app/snaps/keyring-snap-removal-warning/keyring-snap-removal-warning.tsx:160]: hey @georgewrmarshall, if im not wrong we just have to update the `type` prop in files, right? can i work on these changes?

Comment (georgewrmarshall) [reply ui/components/app/snaps/keyring-snap-removal-warning/keyring-snap-removal-warning.tsx:160]: Hey @dhruvv173, thanks so much for getting us this far! I think @garrettbear is going to take it from here as it's a blocker for us but it would be great if you could take on these 2 issues:
- https://github.com/MetaMask/metamask-extension/issues/20155
- https://github.com/MetaMask/metamask-extension/issues/20159
We are super close to migrating 100% of the component-library to TypeScript and the above issues will be contributing to that

**Review by georgewrmarshall — CHANGES_REQUESTED** [review]
There is an introduced `A component is changing an uncontrolled input of type text to be
controlled.` error.

https://github.com/MetaMask/metamask-extension/assets/8112138/926bdbfc-1a02-47e4-b7c6-6463ac126f22

**Review by georgewrmarshall — DISMISSED** [review]
Excellent work! Thanks for your contribution @dhruvv173 and for getting it across the line @garrettbear 🔥

**Review by NidhiKJha — DISMISSED** [review]

**Review by georgewrmarshall — APPROVED** [review]

**Review by montelaidev — APPROVED** [review]
Just tested and the keyring snaps warning is working.

**Review by FrederikBolding — APPROVED** [review]
